TST: Add tests for lapack ?geqrfp

swallan opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Adds tests for #11169
#### What does this implement/fix?
<!--Please explain your changes.-->
This pull request adds tests for the ?geqrpf lapack routine. 

#### Additional information
<!--Any additional information you think is important.-->

swallan committed:
TST: geqrfp; initial add tests for lapack routine

swallan committed:
TST: ?geqrfp test parametrization and tests added
Added parametrization for dtypes, tall, wide, and square matricies. Tests that the routine does not return any negatives on r diagonal even when linalg.qr does. Modifies how q is extracted from the ?geqrfp result for when m>n.

swallan commented:
@mdhaber Opening the pull request so you can see what's been written for this. It should satisfy the parameters from the issue.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+
+    np.random.seed(42)
+    rtol = 250*np.finfo(dtype).eps
+    atol = np.finfo(dtype).eps
```
This might be too stringent. I think (at)ilayn suggested a tolerance on the gttrf/gttrs issue or PR - please take a look for that.

swallan replied:
Got it, I see that. I’ll update this

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+    geqrfp = get_lapack_funcs(('geqrfp'), dtype=dtype)
+    gqr = get_lapack_funcs(("orgqr"), dtype=dtype)
+
+    np.random.seed(42)
```
```suggestion

```
Already done

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+
+    np.random.seed(42)
+    m = matrix_size[0]
+    n = matrix_size[1]
```
m, n = matrix_size

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+def generate_random_dtype_array(shape, dtype):
+    # generates a random matrix of desired data type of shape
+    if type(shape) != tuple:
+        # in case shape is not passed in as a tuple
```
One one hand, I don't think you need this sort of logic in the function. It's not a public function, so I would rely on the user to provide a tuple.
On the other hand, if you were to include this sort of logic, it would probably be better to accept any type of iterable. In this case I would probably use `try-except` (e.g. check for length == 2) rather than `if-else`.
But again, here, I think this should be removed and you should pass in a tuple.

swallan replied:
Got it. I added this so it wouldn’t break gttrf/s since that will need to be updated for how it calls this (it only requests vectors by passing a single n). I'll remove that logic part

mdhaber replied:
Ok, make that pass in (n, n)

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+        # set first n columns of qqr to qr_A
+        qqr[:, :n] = qr_A
+        # determine q from this
+        q = gqr(qqr, tau=tau, lwork=m)[0]
```
Can you add a comment that `m` is sufficient for `lwork` according to LAPACK documentation?

swallan replied:
Yes

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+
+    # obtain q from the orgqr lapack routine:
+
+    if m > n:
```
Can you add a comment explaining where this logic for extracting `q` comes from?

swallan replied:
Yep, it's based on the strategy qr used to grab q with orgqr

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+    assert_allclose(q@r, A, rtol=rtol)
+    # enure that q is orthogonal (that q @ transposed q is the identity)
+    assert_allclose(np.eye(q.shape[0]), q@(q.conj().T), rtol=rtol,
+                    atol=atol*10)
```
```suggestion
                    atol=atol)
```
I'd set `atol` to what you want it to be at the top.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+    assert_allclose(r, np.triu(r), rtol=rtol)
+    # make sure diagonals of r are positive for this random solution
+    assert_(np.all(r[np.arange(r.shape[0]-1),
+                     np.arange(r.shape[0]-1)] > np.zeros(r.shape[0]-1)))
```
Tests are spot on. Great!

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,80 @@ def test_geqrfp_lwork(dtype, shape):
+    # ensure r is upper tri by comparing original r to r as upper triangular
+    assert_allclose(r, np.triu(r), rtol=rtol)
+    # make sure diagonals of r are positive for this random solution
+    assert_(np.all(r[np.arange(r.shape[0]-1),
```
Here and below, use [`np.diag`](https://numpy.org/devdocs/reference/generated/numpy.diag.html?highlight=diag#numpy.diag) to extract the diagonal.

swallan replied:
Great catch

mdhaber reviewed (commented):
Great work - very close! Add a comment as to why we're not doing NAG example for this one. 
Also, I know I didn't include it in the issue, let's check for `info==1` in the case of a normal exit, and let's try to get a negative info (`if INFO = -i, the i-th argument had an illegal value`) by providing some illegal argument (e.g. `None`).

swallan commented:
> Great work - very close! Add a comment as to why we're not doing NAG example for this one.
> Also, I know I didn't include it in the issue, let's check for `info==1` in the case of a normal exit, and let's try to get a negative info (`if INFO = -i, the i-th argument had an illegal value`) by providing some illegal argument (e.g. `None`).


I added this, however, with element 1 illegal (None) it returns 0 with info and then the output factorization is all not a number Nan entries.

swallan committed:
TST: Added test description comment, made tests more succinct.

Changed atol to be 100*.., simplified get_random_matrix function, added clarifying comments.

mdhaber commented:
We don't know what its definition of illegal it's. Not a problem. Maybe try a string?

swallan commented:
@mdhaber Ok, I tried an array with a string in it and an array of just a string, but any string raises an error for casting to a float after the call. I'll think about it more and see if I can put something else in there.

mdhaber commented:
If it doesn't work out in 5 minutes, let's forget about that argument.

swallan committed:
TST: ?geqrfp: Refine comments and remove info test

swallan committed:
TST: ?geqrfp use np.diag instead of indexing

This commit removes long/complicated fancy indexing to determine diagonals and used np.diag(r) instead.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1734,10 +1735,7 @@ def test_geqrfp(dtype, matrix_size):
             np.all(r[np.arange(r.shape[0]-1),
                      np.arange(r.shape[0]-1)] > np.zeros(r.shape[0]-1)))
-    # assert that the first value (1 indexed 1) returned as the illegal value
-    A[0][0] = None
```
I didn't mean to change only one element to `None`. Replace the whole argument with `None`. It might not work because f2py might raise an error rather than actually calling the Fortran routine, but it's worth trying.

swallan replied:
I think I did try that; I double checked and with A=None it runs with output info = 0, and the output factorization is just a single Nan.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,86 @@ def test_geqrfp_lwork(dtype, shape):
+    # assert that any of the entries on the diagonal from linalg.qr
+    #   are negative and that all of geqrfp are positive.
+    assert_(np.any(np.diag(r_rq_neg) < 0) and
+            np.all(np.diag(r) > np.zeros(np.diag(r).shape)))
```
Does
```suggestion
            np.all(np.diag(r) > 0)
```
not work?

swallan replied:
You're right, that does work and is better

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1655,86 @@ def test_geqrfp_lwork(dtype, shape):
+
+    # create random matrix of dimentions m x n
+    A = generate_random_dtype_array((m, n), dtype=dtype)
+    # create rq matrix using geqrfp
```
```suggestion
    # create qr matrix using geqrfp
```

swallan committed:
TST: Fix typo and make tests more succinct

mdhaber reviewed (approved):
@Kai-Striega Please consider for merge! Note: I think `generate_random_dtype_array` is duplicated. How do you want to handle this going forward - rebase/merge so that Sam doesn't copy it in future PRs?

swallan commented:
More on that note–it is duplicated, but the in this pull `generate_random_dtype_array` is improved to take in a tuple instead of just a single number.

Kai-Striega commented:
@swallan I think your test looks good! Just one thing - we should include a test for the NAG examples; [f08aef ](https://www.nag.com/numeric/fl/nagdoc_latest/html/f08/f08aef.html) and [f08asf ](https://www.nag.com/numeric/fl/nagdoc_latest/html/f08/f08asf.html) for real and complex types respectively. 

Regarding ``generate_random_dtype_array`` I can rebase this branch that way Sam will be able to change it.

mdhaber commented:
@Kai-Striega I told @swallan he did not need to do those NAG example tests because they are not as straightforward to implement as usual. I would have expected the NAG example to just perform QR decomposition and give the result, but in this case, it uses the result to solve a least-squares problem.

While the NAG tests are nice to have, I think the random tests are sufficient. Is that OK?

Kai-Striega commented:
@mdhaber OK. That's fine. Regarding the rebase I think it is simpler to merge this PR into the branch and then I'll rebase + fix any merge conflicts from there.

mdhaber commented:
Sounds good!
